File: examples/mock-adapter-python/provider_telemetry.py
"""Provider-to-Runtime telemetry example; no OpenTelemetry SDK is required."""

from __future__ import annotations

import argparse
import json
import uuid
from dataclasses import dataclass

import grpc
from google.protobuf.json_format import ParseDict
from google.protobuf.struct_pb2 import Struct
from google.protobuf.timestamp_pb2 import Timestamp

from generated import provider_telemetry_pb2, provider_telemetry_pb2_grpc


@dataclass(frozen=True)
class TaskTelemetryBinding:
    task_id: str
    external_execution_id: str
    operation_name: str


def json_struct(value: dict[str, object]) -> Struct:
    message = Struct()
    ParseDict(value, message)
    return message


def example_events(binding: TaskTelemetryBinding | None = None):
    occurred_at = Timestamp()
    occurred_at.GetCurrentTime()

    def event(event_type, payload):
        return provider_telemetry_pb2.ProviderTelemetryEvent(
            provider_event_id=str(uuid.uuid4()),
            event_type=event_type,
            resource_id="mock-resource-1",
            resource_type="mock.resource",
            occurred_at=occurred_at,
            attributes=json_struct({"source": "python-example"}),
            payload=json_struct(payload),
        )

    events = [
        event(
            provider_telemetry_pb2.RESOURCE_STATE,
            {"state": "ready", "reasonCode": "RESOURCE_READY"},
        ),
        event(
            provider_telemetry_pb2.RESOURCE_METRIC,
            {"metricName": "utilization", "value": 0.42, "unit": "ratio"},
        ),
        event(
            provider_telemetry_pb2.RESOURCE_HEALTH,
            {"health": "healthy", "reasonCode": "CHECK_OK"},
        ),
    ]
    if binding is not None:
        progress = event(
            provider_telemetry_pb2.EXECUTION_PROGRESS,
            {"current": 4, "total": 10, "percentage": 40, "unit": "items"},
        )
        progress.task_id = binding.task_id
        progress.external_execution_id = binding.external_execution_id
        progress.operation_name = binding.operation_name
        events.append(progress)
    for sequence, item in enumerate(events, start=1):
        item.provider_event_sequence = sequence
    return events


async def emit_with_duplicate_retry(endpoint: str, provider_id: str, events, credentials=None):
    channel = (
        grpc.aio.insecure_channel(endpoint)
        if credentials is None
        else grpc.aio.secure_channel(endpoint, credentials)
    )
    async with channel:
        stub = provider_telemetry_pb2_grpc.ProviderTelemetryIngressStub(channel)
        request = provider_telemetry_pb2.EmitProviderEventsRequest(
            provider_id=provider_id, events=events
        )
        first = await stub.EmitProviderEvents(request)
        duplicate = await stub.EmitProviderEvents(request)
        return first, duplicate


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--self-test", action="store_true")
    arguments = parser.parse_args()
    if arguments.self_test:
        binding = TaskTelemetryBinding("task-example", "execution-example", "durable_task")
        events = example_events(binding)
        print(json.dumps({"eventCount": len(events), "types": [item.event_type for item in events]}))
        return
    raise SystemExit("Use this module from the Adapter or run with --self-test")


if __name__ == "__main__":
    main()

File: examples/mock-adapter-typescript/src/provider-telemetry.ts
import * as grpc from "@grpc/grpc-js";
import { randomUUID } from "node:crypto";
import { telemetryClientConstructor } from "../../../packages/provider-telemetry/src/index.js";

export type ProviderTelemetryEventType =
  "RESOURCE_STATE" | "RESOURCE_METRIC" | "RESOURCE_HEALTH" | "EXECUTION_PROGRESS";

export interface ProviderTelemetryExampleEvent {
  providerEventId: string;
  providerEventSequence: string;
  eventType: ProviderTelemetryEventType;
  resourceId: string;
  resourceType: string;
  taskId?: string;
  externalExecutionId?: string;
  operationName?: string;
  occurredAt: { seconds: string; nanos: number };
  attributes: Record<string, unknown>;
  payload: Record<string, unknown>;
}

export interface TaskTelemetryBinding {
  taskId: string;
  externalExecutionId: string;
  operationName: string;
}

export interface ProviderTelemetryClientOptions {
  endpoint: string;
  providerId: string;
  rootCertificate?: Buffer;
  clientCertificate?: Buffer;
  clientKey?: Buffer;
}

interface ProviderTelemetryGrpcClient extends grpc.Client {
  emitProviderEvents(
    request: Record<string, unknown>,
    callback: (error: grpc.ServiceError | null, response: unknown) => void,
  ): grpc.ClientUnaryCall;
}

export class MockProviderTelemetryClient {
  readonly #providerId: string;
  readonly #client: ProviderTelemetryGrpcClient;

  constructor(options: ProviderTelemetryClientOptions) {
    this.#providerId = options.providerId;
    const Client = telemetryClientConstructor();
    const credentials =
      options.rootCertificate === undefined
        ? grpc.credentials.createInsecure()
        : grpc.credentials.createSsl(
            options.rootCertificate,
            options.clientKey,
            options.clientCertificate,
          );
    this.#client = new Client(
      options.endpoint,
      credentials,
    ) as unknown as ProviderTelemetryGrpcClient;
  }

  emit(events: ProviderTelemetryExampleEvent[]): Promise<Record<string, unknown>> {
    return new Promise((resolve, reject) => {
      this.#client.emitProviderEvents(
        { providerId: this.#providerId, events },
        (error, response) => {
          if (error === null) resolve(response as Record<string, unknown>);
          else reject(error);
        },
      );
    });
  }

  async emitWithDuplicateRetry(
    events: ProviderTelemetryExampleEvent[],
  ): Promise<[Record<string, unknown>, Record<string, unknown>]> {
    const first = await this.emit(events);
    const duplicate = await this.emit(events);
    return [first, duplicate];
  }

  close(): void {
    this.#client.close();
  }
}

export function providerTelemetryExampleEvents(
  binding?: TaskTelemetryBinding,
): ProviderTelemetryExampleEvent[] {
  const occurredAt = timestampNow();
  const resourceId = "mock-resource-1";
  const base = (eventType: ProviderTelemetryEventType, payload: Record<string, unknown>) => ({
    providerEventId: randomUUID(),
    providerEventSequence: "1",
    eventType,
    resourceId,
    resourceType: "mock.resource",
    occurredAt,
    attributes: { source: "typescript-example" },
    payload,
  });
  const events: ProviderTelemetryExampleEvent[] = [
    base("RESOURCE_STATE", { state: "ready", reasonCode: "RESOURCE_READY" }),
    base("RESOURCE_METRIC", { metricName: "utilization", value: 0.42, unit: "ratio" }),
    base("RESOURCE_HEALTH", { health: "healthy", reasonCode: "CHECK_OK" }),
  ];
  if (binding !== undefined) {
    events.push({
      ...base("EXECUTION_PROGRESS", { current: 4, total: 10, percentage: 40, unit: "items" }),
      ...binding,
    });
  }
  return events.map((event, index) => ({ ...event, providerEventSequence: String(index + 1) }));
}

function timestampNow(): { seconds: string; nanos: number } {
  const milliseconds = Date.now();
  return {
    seconds: String(Math.floor(milliseconds / 1_000)),
    nanos: (milliseconds % 1_000) * 1_000_000,
  };
}
